pkg/gds/models/v1/contacts.go

```diff
@@ -0,0 +1,216 @@
+package models
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	TechnicalContact      = "technical"
+	AdministrativeContact = "administrative"
+	LegalContact          = "legal"
+	BillingContact        = "billing"
+)
+
+type contactType struct {
+	contact *pb.Contact
+	kind    string
+	err     error
+}
+
+// Returns True if a Contact is not nil and has an email address.
+func ContactHasEmail(contact *pb.Contact) bool {
+	return contact != nil && contact.Email != ""
+}
+
+// Returns True if a Contact is verified.
+func ContactIsVerified(contact *pb.Contact) (verified bool, err error) {
+	if _, verified, err = GetContactVerification(contact); err != nil {
+		return false, err
+	}
+	return verified, nil
+}
+
+// Returns a standardized order of iterating through contacts.
+func contactOrder(contacts *pb.Contacts) []*contactType {
+	return []*contactType{
+		{contact: contacts.Technical, kind: TechnicalContact},
+		{contact: contacts.Administrative, kind: AdministrativeContact},
+		{contact: contacts.Legal, kind: LegalContact},
+		{contact: contacts.Billing, kind: BillingContact},
+	}
+}
+
+type ContactIterator struct {
+	email    bool
+	verified bool
+	index    int
+	current  *contactType
+	contacts []*contactType
+}
+
+// Returns a new ContactIterator which has Next() and Value() methods that can be used
+// to iterate over contacts in a Contacts object.
+func NewContactIterator(contacts *pb.Contacts, requireEmail bool, requireVerified bool) *ContactIterator {
+	return &ContactIterator{
+		email:    requireEmail,
+		verified: requireVerified,
+		contacts: contactOrder(contacts),
+	}
+}
+
+// Moves the ContactIterator to the next existing contact and returns true if there is
+// one, false if the iteration is complete.
+func (i *ContactIterator) Next() bool {
+	for ; i.index < len(i.contacts); i.index++ {
+		i.current = i.contacts[i.index]
+		contact := i.current.contact
+		if contact != nil {
+			if i.email && !ContactHasEmail(contact) {
+				continue
+			}
+			if i.verified {
+				var verified bool
+				var err error
+				if verified, err = ContactIsVerified(contact); err != nil {
+					// If we can't retrieve the contact verification, let the caller
+					// handle the error on Value().
+					i.current.err = err
+					i.index++
+					return true
+				}
+				if !verified {
+					continue
+				}
+			}
+			i.index++
+			return true
+		}
+	}
+	i.current = nil
+	return false
+}
+
+func (i *ContactIterator) Value() (*pb.Contact, string, error) {
+	if i.current != nil {
+		return i.current.contact, i.current.kind, i.current.err
+	}
+	return nil, "", errors.New("no more contacts")
+}
```

Comment:
I really did like the iterator closure that you put together, but I think you're right that in this case, we need a more standard iterator struct to deal with error handling. This is a good implementation, but I'd like to recommend that we use recursion here in case we extend our filtering context. I'd also like to suggest we reduce some of our data storage by using the references we already have and we implement a multi-error handling functionality. 

The one big note is that for this recursive implementation to work, the index must be initialized to -1. 

```suggestion
type ContactIterator struct {
	email    bool
	verified bool
	index    int
	contacts []*contactType
}

// Returns a new ContactIterator which has Next() and Value() methods that can be used
// to iterate over contacts in a Contacts object.
func NewContactIterator(contacts *pb.Contacts, requireEmail bool, requireVerified bool) *ContactIterator {
	return &ContactIterator{
		email:    requireEmail,
		verified: requireVerified,
		index: -1,
		contacts: contactOrder(contacts),
	}
}

// Moves the ContactIterator to the next existing contact and returns true if there is
// one, false if the iteration is complete.
func (i *ContactIterator) Next() bool {
	// Advance the index (note, index must be initialized at -1) 
	i.index++
	
	// Stopping condition
	if i.index >= len(i.contacts) {
		return false
	}
	
	// Filter checks - contact must exist
	contact := i.contacts[i.index].contact
	if contact == nil {
		return i.Next()
	}
	
	// Filter if email required
	if i.email && !ContactHasEmail(contact) {
		return i.Next()
	}
	
	// Filter if verified is required
	if i.verified {
		if verified, err := ContactIsVerified(contact); err != nil || !verified {
			// Even in an error we're skipping the contact, errors have to be 
			// fetched as a multi-error after the iteration is complete.
			contact.err = err 
			return i.Next()
		}
	}
	
	return true
}

func (i *ContactIterator) Value() (*pb.Contact, string) {
	if i.index < len(i.contacts) {
		// Note that no checking of the contact occurs here to allow
		// us to allow iteration with different filtering mechanisms.
		current := i.contacts[i.index]
		return current.contact, current.kind
	}
	return nil, ""
}

// Error returns a multi-error that describes any errors that occurred during
// iteration for any contact. This will allow us to log all errors once.
func (i *ContactIterator) Error() error {
	bad := make([]string, 0, 4)
	for _, contact := range i.contacts {
		if contact.err != nil {
			bad = append(bad, contact.type)
		}
	}
	if len(bad) > 0 {
		bads := strings.Join(contacts, ", ")
		return fmt.Errorf("errors occurred processing the %s contacts", bads)
	}
	return nil
}
```

Reply:
Note that we could actually use a multi-error if you wanted to track all errors across all contacts: https://github.com/hashicorp/go-multierror